adapters/repos/db/roaringsetrange/segment_in_memory.go

```diff
@@ -0,0 +1,241 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package roaringsetrange
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/weaviate/sroar"
+	"github.com/weaviate/weaviate/adapters/repos/db/roaringset"
+	"github.com/weaviate/weaviate/entities/concurrency"
+	"github.com/weaviate/weaviate/entities/filters"
+)
+
+type SegmentInMemory struct {
+	lock    *sync.RWMutex
+	bitmaps rangeBitmaps
+}
+
+func NewSegmentInMemory() *SegmentInMemory {
+	s := &SegmentInMemory{
+		lock: new(sync.RWMutex),
+	}
+	for key := range s.bitmaps {
+		s.bitmaps[key] = sroar.NewBitmap()
+	}
+	return s
+}
+
+func (s *SegmentInMemory) MergeSegmentByCursor(cursor SegmentCursor) error {
+	key, layer, ok := cursor.First()
+	if !ok {
+		// empty segment, nothing to merge
+		return nil
+	}
+	if key != 0 {
+		return fmt.Errorf("invalid first key of merged segment")
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if deletions := layer.Deletions; !deletions.IsEmpty() {
+		for key := range s.bitmaps {
+			s.bitmaps[key].AndNotConc(deletions, concurrency.SROAR_MERGE)
+		}
+	}
+	for ; ok; key, layer, ok = cursor.Next() {
+		s.bitmaps[key].OrConc(layer.Additions, concurrency.SROAR_MERGE)
+	}
+	return nil
+}
+
+func (s *SegmentInMemory) MergeMemtable(memtable *Memtable) error {
+	nodes := memtable.Nodes()
+	if len(nodes) == 0 {
+		// empty memtable, nothing to merge
+		return nil
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if deletions := nodes[0].Deletions; !deletions.IsEmpty() {
+		for key := range s.bitmaps {
+			s.bitmaps[key].AndNotConc(deletions, concurrency.SROAR_MERGE)
+		}
+	}
+	for _, node := range nodes {
+		s.bitmaps[node.Key].OrConc(node.Additions, concurrency.SROAR_MERGE)
```

Comment:
is `s.bitmaps[0]` holding the entire set of docIDs ? nodes[0].Deletions seems to have the entire set of deleted docIDs while nodes[0].Additions seems to have the entire set of docIDs

if it were that the case then we may get duplicated docIDs while reading?

Reply:
`s.bitmaps[0]` is non-null bitmap. it holds all the docIds that are indexed. bitmaps 1-64 hold values split by bits.
therefore document with prop of value = 0 will be added only to bitmaps[0] (no bits) and document with prop of value = 5 will be added to bitmaps[0], bitmaps[1] and bitmaps[3].

memtable also have non-null node. the one with index 0. it holds docIds added (Additions) and deleted (Deletions). Since docId can be assigned just to one value (that effectively replaces previous one) docIds that were added to Additions are also added to Deletions (to invalidate previous values in earlier segments, here in in-memory segment). deletions does not have to split to bits, therefore only node0 stores deletions bitmap